File: 0036-valid-sudoku/0036-valid-sudoku.java
class Solution {
    int mask = 0;
    
    public boolean isValidSudoku(char[][] board) {
        boolean result = true;
        for(int i = 0; result && i < board.length; i++) result &= checkRow(board, i);
        for(int i = 0; result && i < board[0].length; i++) result &= checkCol(board, i);
        for(int i = 0; result && i < board.length; i += 3) {
            for(int j = 0; result && j < board[0].length; j += 3) {
                result &= checkBox(board, new int[] {i, j}, new int[] {i + 2, j + 2});
            }
        }
        
        return result;
    }
    
    boolean checkRow(char[][] board, int row) {
        reset();
        for(int c = 0; c < board[row].length; c++) {
            if (!add(board[row][c])) return false;
        }
        return true;
    }
    
    boolean checkCol(char[][] board, int col) {
        reset();
        for(int r = 0; r < board.length; r++) {
            if (!add(board[r][col])) return false;
        }
        return true;
    }
    
    boolean checkBox(char[][] board, int[] ul, int[] br) {
        reset();
        for(int r = ul[0]; r <= br[0]; r++) {
            for(int c = ul[1]; c <= br[1]; c++) {
                if (!add(board[r][c])) return false;
            }
        }
        return true;
    }
    
    void reset() {
        mask = 0;
    }
    
    boolean add(char n) {
        if (n == '.') return true;
        
        int b = 1 << (Character.getNumericValue(n) - 1);
        if ((mask & b) != 0) return false;
        mask |= b;
        
        return true;
    }
}

File: 0036-valid-sudoku/0036-valid-sudoku.py
class Solution(object):
    def isValidSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: bool
        """
        for i in range(9):
            if (not self.checkCol(board, i) or not self.checkRow(board, i)):
                return False
            
        for i in range(3):
            r = i * 3
            for j in range(3):
                c = j * 3
                if not self.checkBox(board, [r, c], [r + 2, c + 2]):
                    return False
                
        return True
        
        
    def checkCol(self, board, col):
        seen = [False] * 10
        for i in range(9):
            if board[i][col] == '.':
                continue
            if seen[int(board[i][col])]:
                return False
            seen[int(board[i][col])] = True
        return True
            
    def checkRow(self, board, row):
        seen = [False] * 10
        for i in range(9):
            if board[row][i] == '.':
                continue
            if seen[int(board[row][i])]:
                return False
            seen[int(board[row][i])] = True
        return True
        
        
    def checkBox(self, board, ul, br):
        seen = [False] * 10
        for r in range(ul[0], br[0] + 1):
            for c in range(ul[1], br[1] + 1):
                if board[r][c] == '.':
                    continue
                if seen[int(board[r][c])]:
                    return False
                seen[int(board[r][c])] = True
        return True